contracts/token/ERC1155/ERC1155Base.sol

```diff
@@ -236,6 +204,31 @@ abstract contract ERC1155Base is IERC1155, ERC165 {
     }
   }
 
+  function _transferBatch (
+    address operator,
+    address sender,
+    address recipient,
+    uint[] memory ids,
+    uint[] memory amounts,
+    bytes memory data
+  ) virtual internal {
+    require(recipient != address(0), 'ERC1155: transfer to the zero address');
+    require(sender == msg.sender || isApprovedForAll(sender, msg.sender), 'ERC1155: caller is not owner nor approved');
+
+    _beforeTokenTransfer(operator, sender, recipient, ids, amounts, data);
+    _doSafeBatchTransferAcceptanceCheck(operator, sender, recipient, ids, amounts, data);
+
+    mapping (uint => mapping (address => uint)) storage balances = ERC1155BaseStorage.layout().balances;
+    for (uint i=0; i < ids.length; i++) {
+      uint token = ids[i];
+      uint amount = amounts[i];
+      balances[token][sender] -= amount;
```

Comment:
Replace the following comments:
```
// TODO: error message
// balances[id][from] = balances[id][from].sub(amount, 'ERC1155: insufficient balances for transfer');
```

I still need to decide whether to use the default underflow message, or throw a custom error.